cider/exec/operator/join/CiderLinearProbingHashTable.h

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+A hash map for join. Uses open addressing with linear probing.
+Advantages:
+  - Predictable performance. Doesn't use the allocator unless load factor
+    grows beyond 50%. Linear probing ensures cash efficency.
+  - Desgin for no delete/erase action, makes it faster on insert and lookup
+  - Allow duplicate keys
+Disadvantages:
+  - Significant performance degradation at high load factors.
+  - Maximum load factor hard coded to 50%, memory inefficient.
+ */
+#pragma once
+
+#include <folly/FBVector.h>
+#include <algorithm>
+#include <cstddef>
+#include <cstdint>
+#include <limits>
+#include <stdexcept>
+#include <vector>
+
+namespace cider_ht {
+template <typename KT>
+struct table_key {
+  KT key;
+  bool is_not_null;
+  std::size_t duplicate_num;
+};
+
+// TODO: extends JoinHashTable class and a hashtable basic interface
+template <typename Key,
+          typename T,
+          typename Hash = std::hash<Key>,
+          typename KeyEqual = std::equal_to<void>,
+          typename Allocator = std::allocator<std::pair<table_key<Key>, T>>>
+class LPHashTable {
+ public:
+  using key_type = table_key<Key>;
+  using mapped_type = T;
+  using value_type = std::pair<key_type, T>;
+  using size_type = std::size_t;
+  using hasher = Hash;
+  using key_equal = KeyEqual;
+  using allocator_type = Allocator;
+  using reference = value_type&;
+  using buckets = folly::fbvector<value_type, allocator_type>;
+
+  template <typename ContT, typename IterVal>
+  struct hm_iterator {
+    using difference_type = std::ptrdiff_t;
+    using value_type = IterVal;
```

Comment:
Seems `value_type` and `reference` are all named before. What's the difference? maybe better avoiding using these alias which causes confusion?

Reply:
reference is value_type& and pointer is value_type*. This seems is a common code tyle in other hashmap. For example
:
1. https://github.com/abseil/abseil-cpp/blob/master/absl/container/internal/raw_hash_set.h#L1153

2. https://github.com/facebook/folly/blob/main/folly/container/F14Map.h#L101